go-controller/pkg/cni/helper_linux.go

```diff
@@ -252,81 +254,100 @@ func setupInterface(netns ns.NetNS, containerID, ifName string, ifInfo *PodInter
 	return hostIface, contIface, nil
 }
 
+var chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890-"
+
+func generateId(length int) string {
+	charsLength := len(chars)
+	b := make([]byte, length)
+	_, err := rand.Read(b) // generates len(b) random bytes
+	if err != nil {
+		klog.Errorf("can't generate a random ID: ", err)
+		return ""
+	}
+
+	for i := 0; i < length; i++ {
+		b[i] = chars[int(b[i])%charsLength]
+	}
+	return string(b)
+}
+
+// generate a unique interface name for the temporary netdev that will be moved to pod namespace
+func generateIfName(containerID string) string {
+	randomId := generateId(5) // random ID with 5 chars
+	// ifname max length is 15
+	return containerID[:(15-len(randomId))] + randomId
+}
+
 // Setup sriov interface in the pod
 func setupSriovInterface(netns ns.NetNS, containerID, ifName string, ifInfo *PodInterfaceInfo, deviceID string) (*current.Interface, *current.Interface, error) {
 	hostIface := &current.Interface{}
 	contIface := &current.Interface{}
-	ifnameSuffix := ""
-
 	netdevice := ifInfo.NetdevName
 
 	// 1. Move netdevice to Container namespace
-	err := moveIfToNetns(netdevice, netns)
-	if err != nil {
-		return nil, nil, err
-	}
-
-	err = netns.Do(func(hostNS ns.NetNS) error {
-		contIface.Name = ifName
-		err = renameLink(netdevice, contIface.Name)
+	if len(netdevice) != 0 {
+		// for avoiding conflicts, the netdevice is rename to a unique ifname before moving to the pod namespace
+		tempNetdevice := generateIfName(containerID)
+		err := renameLink(netdevice, tempNetdevice)
 		if err != nil {
-			return err
-		}
-		link, err := util.GetNetLinkOps().LinkByName(contIface.Name)
-		if err != nil {
-			return err
-		}
-		err = util.GetNetLinkOps().LinkSetHardwareAddr(link, ifInfo.MAC)
-		if err != nil {
-			return err
-		}
-		err = util.GetNetLinkOps().LinkSetMTU(link, ifInfo.MTU)
-		if err != nil {
-			return err
+			return nil, nil, err
 		}
-		err = util.GetNetLinkOps().LinkSetUp(link)
+		err = moveIfToNetns(tempNetdevice, netns)
```

Comment:
I think we can avoid renaming for every pod by:
1. call `moveIfToNetns` directly without renaming, for most cases it will just go through
2. if `moveIfToNetns` unfortunately returns `file exists` error, go ahead to rename and set the new name in `ifInfo.NetdevName`, then return error
3. CNI should retry with the new random name

Reply:
I hope I've done what you suggested, thanks!

Reply:
yeah, LGTM